src/plugins/discover/public/application/helpers/get_sharing_data.ts

```diff
@@ -66,20 +33,45 @@ export async function getSharingData(
   searchSource.removeField('aggs');
   searchSource.removeField('size');
 
-  const body = await searchSource.getSearchRequestBody();
+  // Set the fields of the search source to match the saved search columns
+  searchSource.removeField('fields');
+  searchSource.removeField('fieldsFromSource');
+
+  let columns = state.columns || [];
+
+  // NOTE: A newly saved search with no columns selected has a bug(?) where the
+  // column array is a single '_source' value which is invalid for CSV export
+  if (columns && columns.length === 1 && /^_source$/.test(columns.join())) {
+    columns = [];
+  }
+
+  // conditionally add the time field column
+  let timeFieldName: string | undefined;
+  const hideTimeColumn = config.get(DOC_HIDE_TIME_COLUMN_SETTING);
+  if (!hideTimeColumn && index && index.timeFieldName) {
+    timeFieldName = index.timeFieldName;
+  }
+
+  if (columns && columns.length > 0 && timeFieldName) {
+    columns = [timeFieldName, ...columns];
+  }
+
+  if (columns.length === 0) {
+    searchSource.setField('fields', ['*']);
```

Comment:
`searchSource.getField('fieldsFromSource') ? searchSource.setField('fieldsFromSource', ['*']) : searchSource.setField('fields', ['*']);`

Reply:
In this code I have tried to abstract away `fieldsFromSource` and rely on `fields`.  I assumed it would be OK, and I still don't really understand the downside of doing it this way. I need a test case to check against.

Whether I have Kibana set to `discover:searchFieldsFromSource` or not, I seem to get all the CSV columns I expect - whether or not I select fields as columns in my saved search.